documents/Specification/MaterialX.NPRSpec.md

```diff
@@ -0,0 +1,105 @@
+<!-----
+MaterialX NPR Shading Nodes v1.39
+----->
+
+
+# MaterialX NPR Shading Nodes
+
+**Version 1.39**
+Doug Smythe - Industrial Light & Magic
+Jonathan Stone - Lucasfilm Advanced Development Group
+April 2, 2024
+
+# Introduction
+
+The MaterialX Specification and MaterialX Physically Based Shading Nodes documents describe a number of standard pattern and shading nodes that may be used to construct nodegraph-based shaders for physically based rendering in a variety of applications.  However, there are certain operations that are desirable in non-photorealistic shading styles but which cannot be implemented within certain rendering constructs.  It is also helpful conceptually to separate nodes primarily useful for photorealistic and non-photorealistic shading styles into separate libraries.
+
+This document describes a number of MaterialX nodes primarily applicable to non-photorealistic, or NPR, rendering.  Rendering applications whose architecture cannot support these operations are not required to support these nodes.
+
+
+## Table of Contents
+
+**[MaterialX NPR Library](#materialx-npr-library)**  
+ [NPR Application Nodes](#npr-application-nodes)  
+ [NPR Utility Nodes](#npr-utility-nodes)  
+ [NPR Light Nodes](#npr-light-nodes)  
+ [NPR Shading Nodes](#npr-shading-nodes)  
+
+
+
+## MaterialX NPR Library
+
+
+## NPR Application Nodes
+
+<a id="node-viewdirection"> </a>
+
+* **`viewdirection`**: the current scene view direction (e.g. from the viewing/camera position to the current shading position).  If `viewdirection` is used in a PBR shading context, it should be noted that this would be the same as the incident ray direction for primary ("camera") rays but **not** for secondary/reflection rays.  This node must be of type vector3.
+
+    * `space` (uniform string):  the space in which to return the view vector direction, defaults to "world". 
+
+<a id="node-lightdirection"> </a>
+
+* **`lightdirection`**: the predominant incoming light direction, as defined by the shading environment.  This node must be of type vector3.
+
+    * `space` (uniform string):  the space in which to return the lighting vector direction, defaults to "world". 
+
+
+## NPR Utility Nodes
+
+<a id="node-facingratio"> </a>
+
+* **`facingratio`**: returns the geometric facing ratio, computed as the dot product between the view direction and geometric normal.  Output is a float between 0.0 and 1.0.
+
+    * `viewdirection` (vector3): the viewing direction, defaults to the value of the "Vworld" (world space view direction) geometric property.
+    * `normal` (vector3): the surface normal vector, defaults to the value of the "Nworld" (world space view direction) geometric property.  This vector is expected to be prenormalized to length 1.0.
+    * `faceforward` (boolean): description needed; default is false.
+    * `invert` (boolean): description needed; default is false.
+
+
+## NPR Light Nodes
```

Comment:
Maybe we should omit these lighting nodes from the specification, since they're just implementation details of our GLSL/MSL/ESSL shader generators?  I wouldn't expect artists to have access to these nodes in an editor, for example.

Reply:
Yeah, I wasn't 100% sure these should be added either, though I thought I'd at least put something in to start a discussion.
Similarly, I should probably remove the PBR "environment" light node, which hasn't been implemented and I'm actually not sure it's useful since one would never connect a light node into a material network (right?).

Reply:
This is done in the updated PR